deepsubtree.go

```diff
@@ -128,7 +251,160 @@ func (dst *DeepSubTree) recursiveSet(node *Node, key []byte, value []byte) (
 	default:
 		return nil, false, fmt.Errorf("inner node does not have key set correctly")
 	}
-	return node, updated, nil
+	if updated {
+		return node, updated, nil
+	}
+	err = node.calcHeightAndSize(dst.ImmutableTree)
+	if err != nil {
+		return nil, false, err
+	}
+	orphans := dst.prepareOrphansSlice()
+	node.persisted = false
+	newNode, err := dst.balance(node, &orphans)
+	node.persisted = true
+	if err != nil {
+		return nil, false, err
+	}
+	return newNode, updated, err
+}
+
+// remove tries to remove a key from the tree and if removed, returns its
+// value, nodes orphaned and 'true'.
+func (dst *DeepSubTree) Remove(key []byte) (value []byte, removed bool, err error) {
+	if dst.root == nil {
+		return nil, false, nil
+	}
+	newRootHash, newRoot, value, err := dst.recursiveRemove(dst.root, key)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !dst.skipFastStorageUpgrade {
+		dst.addUnsavedRemoval(key)
+	}
+
+	if newRoot == nil && newRootHash != nil {
+		dst.root, err = dst.ndb.GetNode(newRootHash)
+		if err != nil {
+			return nil, false, err
+		}
+	} else {
+		dst.root = newRoot
+	}
```

Comment:
nit: I love saving loc :-)

```suggestion
	if newRoot == nil && newRootHash != nil {
		newRoot, err = dst.ndb.GetNode(newRootHash)
		if err != nil {
			return nil, false, err
		}
	}
	dst.root = newRoot
```